libsolidity/lsp/ReferenceCollector.cpp

```diff
@@ -0,0 +1,106 @@
+/*
+	This file is part of solidity.
+
+	solidity is free software: you can redistribute it and/or modify
+	it under the terms of the GNU General Public License as published by
+	the Free Software Foundation, either version 3 of the License, or
+	(at your option) any later version.
+
+	solidity is distributed in the hope that it will be useful,
+	but WITHOUT ANY WARRANTY; without even the implied warranty of
+	MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+	GNU General Public License for more details.
+
+	You should have received a copy of the GNU General Public License
+	along with solidity.  If not, see <http://www.gnu.org/licenses/>.
+*/
+// SPDX-License-Identifier: GPL-3.0
+#include <libsolidity/lsp/ReferenceCollector.h>
+
+#include <libsolidity/ast/AST.h>
+#include <libsolidity/lsp/LanguageServer.h>
+
+using namespace solidity::frontend;
+using namespace std::string_literals;
+using namespace std;
+
+namespace solidity::lsp
+{
+
+ReferenceCollector::ReferenceCollector(
+	frontend::Declaration const& _declaration,
+	std::string const& _sourceIdentifierName
+):
+	m_declaration{_declaration},
+	m_sourceIdentifierName{_sourceIdentifierName.empty() ? _declaration.name() : _sourceIdentifierName}
+{
+}
+
+std::vector<DocumentHighlight> ReferenceCollector::collect(
+	frontend::Declaration const* _declaration,
+	frontend::ASTNode const& _ast,
+	std::string const& _sourceIdentifierName
+)
+{
+	if (!_declaration)
+		return {};
+
+	// TODO if vardecl, just use decl's scope (for lower overhead).
+	ReferenceCollector collector(*_declaration, _sourceIdentifierName);
+	_ast.accept(collector);
+	return move(collector.m_result);
+}
+
+void ReferenceCollector::endVisit(frontend::ImportDirective const& _import)
+{
+	for (auto const& symbolAlias: _import.symbolAliases())
+		if (m_sourceIdentifierName == *symbolAlias.alias)
+		{
+			m_result.emplace_back(DocumentHighlight{symbolAlias.location, DocumentHighlightKind::Text});
```

Comment:
If they are all `DocumentHighlightKind::Text`, does it make sense to first only return a vector of locations and then turn them into a vector of DecumentHighlight?